official/mnist/mnist_multi_gpu.py

```diff
@@ -0,0 +1,80 @@
+#  Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Allows the MNIST model to run multi-gpu.
+
+TODO(karmel): When multi-GPU is out of contrib, use core instead.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import sys
+
+import mnist
+import tensorflow as tf
+from tensorflow.python.client import device_lib
+
+
+def model_fn_with_tower_optimizer(features, labels, mode, params):
```

Comment:
Instead of a separate driving file (and the resulting refactoring of `mnist.py`), would it make sense to merge this into `mnist.py`? Some ideas off the top of my head:

- Add a `experimental_use_multiple_gpus` flag to `mnist.py` (which defaults to `False`). With the default value, no `contrib` APIs will be invoked.
- When that flag value is `True`, then we wrap `model_fn` with `tf.contrib.estimator.replicated_model_fn`
- And we also translate that flag to a value in the `params` dictionary to `model_fn`, so that the `TowerOptimizer` wrapping only happens when that param is set.

My gut feeling is that this will reduce the amount of code overall and also make what changes between single and multi GPU to the reader - while avoiding the use of `contrib` APIs by default.

Thoughts?

Reply:
Maybe better discussed offline, but I tend to prefer the separation. Pros:

* Separates out contrib imports without introducing logic/conditionals.
* This puts the responsibility on the main loop to be reusable by other loops, whereas one file would push us towards the world where each run flavor introduces a bunch of conditionals into the main loop.
* This puts the responsibility on the multi-gpu loop to not affect the default run.
* This makes it more obvious to users what is doing what. Someone can use mnist.py as an example for running on single GPU without copying over a bunch of code that is useless to them, and someone looking for an example of how to run multi-GPU will find it much easier to understand exactly the set of changes they will need to make in their own models.
* In the world where there are many cooks in the Model Garden kitchen, more modules is easier than one big module.

(If I have not convinced you, let's discuss offline to avoid the inevitable back and forth.)